src/basic/khash.c

```diff
@@ -43,6 +43,66 @@ struct khash {
         bool digest_valid;
 };
 
+int khash_supported(void) {
+        static const union {
+                struct sockaddr sa;
+                struct sockaddr_alg alg;
+        } sa = {
+                .alg.salg_family = AF_ALG,
+                .alg.salg_type = "hash",
+                .alg.salg_name = "sha256", /* a very common algorithm */
+        };
+
+        static int cached = -1;
```

Comment:
I'm not sure about the cache. If modules are used and they are loaded after `khash_supported` is called, `systemd` will have to be reexecuted manually.

Reply:
modules are loaded on demand when an algorithm is requested, hence kash_supported() should load sha256 if it wasn't loaded yet and all is good.

Reply:
Modules can appear on a machine after `kash_supported` is called for the first time. In this case autoloading doesn't work because there is nothing to load.

Reply:
Well, sure, a lot of things can happen. But first of all, I doubt we need to support that kind of setup too much, and more importantly the reason the code checks for *sha256* of all digests explicitly is that that's such a common digest that it simply makes no sense to compile it as a module. Or to say this differently: if you actually compile your kernel like that and package sha256 as such a "late" module, then it's your own fault, and you get to keep the pieces. I mean, this is *not* an auxiliary module, it's pretty core for many things, hence it's fine to use as a test I think.

Note that nothing "new" really breaks here anyway. all that happens is that failure during early boot will also mean failure during late boot – until reexec. It's not that something that didn't fail before would suddenly fail. I just means it fails for a bit longer. And I think that's a good thing, because people should not expect this to work. Either they should accept the failure for all time or should fix the failure for all time, but accepting the failure during early boot but not during late boot is just strange.

Reply:
I'm sorry. I seem to have mistaken `khash_new` for something else that is widely used by `systemd`. Given that `khash_new`, `khash_new_with_key` and `sd_id128_get_machine_app_specific` are only called by the tests, nothing is really going to break. I was too quick to panic :-)